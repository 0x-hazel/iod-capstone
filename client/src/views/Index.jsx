import { useMutation, useQuery } from "@tanstack/react-query";
import axios from "axios";
import { useNavigate } from "react-router";

export default function Index() {
    const navigate = useNavigate();
    const status = useQuery({
        queryKey: ['status'],
        queryFn: async () => {
            return axios.get("/api/auth/status");
        }
    });
    const logOut = useMutation({
        mutationFn: async () => {
            return axios.post("/api/auth/logout");
        },
        onSuccess: () => {
            navigate(0);
        }
    })
    const loggedIn = (!status.isPending) && (status.data.data.status != "failed");
    return (
        <>
            <p>Hello, world!</p>
            { loggedIn ? <><p>Logged In</p><button className="btn" onClick={logOut.mutate}>Log Out</button></> : <p>Not Logged In</p> }
        </>
    );
}