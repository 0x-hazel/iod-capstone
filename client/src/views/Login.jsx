import { useMutation } from "@tanstack/react-query";
import axios from "axios";
import qs from "qs";
import React from "react";
import { useNavigate } from "react-router";
import { useAlert } from "../hooks/alerts";
import NavBar from "../components/navBar";

export default function Login() {
    const navigate = useNavigate();
    const alert = useAlert();
    const mutation = useMutation({
        mutationFn: (event) => {
            event.preventDefault();
            const formData = new FormData(event.target);
            return axios.post("/api/auth/login", qs.stringify(Object.fromEntries(formData.entries())));
        },
        onSuccess: (response) => {
            if (response.status == 200) {
                navigate("/");
            }
        },
        onError: (result) => {
            alert({status: "error", message: `Error ${result.status}: ${result.response.data.message}`});
        }
    });
    return (
        <>
            <NavBar />
            <div className="flex items-center justify-center h-screen">
                <div className="card card-border bg-base-100 w-96">
                    <div className="card-body items-center text-center">
                        <h2 className="card-title mb-4">Log In</h2>
                        <form onSubmit={mutation.mutate} className="w-full">
                            <label className="floating-label mb-8 w-full">
                                <span>Your Username</span>
                                <input type="text" name="username" className="input validator" required placeholder="Username" />
                            </label>
                            <label className="floating-label mb-8 w-full">
                                <span>Your Password</span>
                                <input type="password" name="password" className="input" required placeholder="Password" minLength={6} />
                            </label>
                            <div className="card-actions">
                                <button className="btn btn-primary">Log In</button>
                            </div>
                        </form>
                    </div>
                </div>
            </div>
        </>
    )
}