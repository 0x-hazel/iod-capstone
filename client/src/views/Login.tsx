import React from "react";

export default function Login() {
    return (
        <div className="flex items-center justify-center h-screen">
            <div className="card card-border bg-base-100 w-96">
                <div className="card-body items-center text-center">
                    <h2 className="card-title mb-4">Log In</h2>
                    <label className="floating-label mb-8 w-full">
                        <span>Your Username</span>
                        <input type="text" className="input validator" required placeholder="Username" />
                    </label>
                    <label className="floating-label mb-8 w-full">
                        <span>Your Password</span>
                        <input type="password" className="input" required placeholder="Password" minLength={6} />
                    </label>
                    <div className="card-actions">
                        <button className="btn btn-primary">Log In</button>
                    </div>
                </div>
            </div>
        </div>
    )
}