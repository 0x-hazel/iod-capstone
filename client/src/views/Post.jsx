import { useQuery } from "@tanstack/react-query";
import axios from "axios";
import { useParams } from "react-router";
import MarkdownView from "../components/markdownView";

export default function Post() {
    const {user, post} = useParams();
    const contentQuery = useQuery({
        queryKey: ['post'],
        queryFn: () => {
            return axios.get(`/api/post/get-post/${user}/${post}`);
        }
    });
    if (contentQuery.isLoading) {
        return <p>Loading...</p>;
    }
    const data = contentQuery.data.data;
    return (
        <div className="flex items-center justify-center w-screen flex-col">
            <div className="p-4 w-full md:w-11/17">
                <div className="mb-4 md:mb-8">
                    <h1 className="text-lg my-2 md:my-4 mx-8">{data.title}</h1>
                    <MarkdownView contents={data.contents} />
                </div>
            </div>
        </div>
    )
}