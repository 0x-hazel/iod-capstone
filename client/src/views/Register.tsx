import { useMutation } from "@tanstack/react-query";
import axios from "axios";
import React, { useMemo, useState } from "react";
import { useNavigate } from "react-router";
import escapeStringRegexp from 'escape-string-regexp';
import qs from "qs";

export default function Register() {
    const navigate = useNavigate();
    const mutation = useMutation({
        mutationFn: (event: Event) => {
            event.preventDefault();
            const formData = new FormData(event.target as HTMLFormElement);
            return axios.post("/api/auth/create-account", qs.stringify(Object.fromEntries(formData.entries())));
        },
        onSuccess: (response) => {
            if (response.status == 200) {
                navigate("/");
            }
        }
    });
    const [password, setPassword] = useState("");
    const passwordPattern = useMemo(() => escapeStringRegexp(password), [password]);
    return (
        <div className="flex items-center justify-center h-screen">
            <div className="card card-border bg-base-100 w-96">
                <div className="card-body items-center text-center">
                    <h2 className="card-title mb-8">Register</h2>
                    <form onSubmit={mutation.mutate} className="w-full">
                        <label className="floating-label w-full h-24">
                            <span>Your Username</span>
                            <input type="text" id="username" name="username" className="input validator" required placeholder="Username"
                                pattern="[A-Za-z][A-Za-z0-9\-]*" minlength="3" maxlength="30" title="Only letters, numbers or dash" />
                            <p className="validator-hint">
                                Must be 3 to 30 characters containing only letters, numbers or dash.
                            </p>
                        </label>
                        <label className="floating-label w-full h-24">
                            <span>Your Password</span>
                            <input type="password" id="password" name="password" className="input validator" required placeholder="Password" minlength="6"
                                pattern="(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{5,}"
                                title="Must be at least 6 characters, including number, lowercase letter, uppercase letter"
                                onChange={e => setPassword(e.target.value)} value={password} />
                            <p className="validator-hint">
                                Must be at least 6 characters, including at least one number, at least one lowercase letter, and at least one uppercase letter.
                            </p>
                        </label>
                        <label className="floating-label w-full h-24">
                            <span>Your Password Again</span>
                            <input type="password" id="password2" name="password2" className="input validator" required placeholder="Confirm Password" minlength="6"
                                pattern={passwordPattern}
                                title="Must match the previous password" />
                            <p className="validator-hint">
                                Must match the previous password.
                            </p>
                        </label>
                        <div className="card-actions">
                            <button className="btn btn-primary">Register</button>
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}