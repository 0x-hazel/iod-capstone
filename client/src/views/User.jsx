import { useMutation, useQuery } from "@tanstack/react-query";
import axios from "axios";
import { useNavigate, useParams } from "react-router";
import CompactPostView from "../components/compactPostView";
import NavBar from "../components/navBar";
import WriteButton from "../components/writeButton";
import EditableAvatar from "../components/editableAvatar";
import EditableTextLine from "../components/editableTextLine";
import EditableTextField from "../components/editableTextField";
import { useCallback, useState } from "react";
import { useAlert } from "../hooks/alerts";

export default function User() {
    const navigate = useNavigate();
    const { user } = useParams();
    const alert = useAlert();
    const contentQuery = useQuery({
        queryKey: ['author'],
        queryFn: () => {
            return axios.get(`/api/user/get-user/${user}`);
        }
    });
    const [isEdited, setIsEdited] = useState(false);
    const mutateData = useMutation({
        mutationFn: (event) => {
            event.preventDefault();
            const formData = new FormData(event.target);
            return axios.post(`/api/user/update-user/${user}`, formData);
        },
        onSuccess: () => {
            setIsEdited(false);
            navigate(0);
        },
        onError: (result) => {
            alert({ status: "error", message: `Error ${result.status}: ${result.response.data.message}` });
        }
    })
    const [avatar, setAvatar] = useState("");
    const [displayName, setDisplayName] = useState("");
    const [bio, setBio] = useState("");
    const setEdited = useCallback(() => setIsEdited(true), [setIsEdited]);
    if (contentQuery.isLoading) {
        return <p>Loading...</p>;
    }
    const data = contentQuery.data.data;
    return (
        <>
            <NavBar />
            <div className="flex items-center justify-center w-screen flex-col">
                <div className="p-4 w-full md:w-11/17">
                    <div className="mb-4 md:mb-8">
                        <div className="card card-border border-secondary w-full mb-4 md:mb-8">
                            <div className="card-body">
                                <div className="flex flex-row mb-[1rem]">
                                    {data.can_edit ?
                                        <EditableAvatar image={data.author.avatar} onEdit={setEdited} updateValue={setAvatar} /> :
                                        <div className="avatar">
                                            <div className="w-24 rounded-full">
                                                <img src={data.author.avatar || "/icon.svg"} />
                                            </div>
                                        </div>
                                    }
                                    <div className="ml-4 md:ml-8 flex flex-col justify-center">
                                        {data.can_edit ?
                                            <EditableTextLine
                                                defaultContent={data.author.display_name}
                                                className="text-2xl font-semibold"
                                                placeholder="Display Name"
                                                onEdit={setEdited}
                                                updateValue={setDisplayName} /> :
                                            <h1 className="text-2xl font-semibold">{data.author.display_name}</h1>
                                        }
                                        <h2 className="text-xl">@{data.author.username}</h2>
                                    </div>
                                </div>
                                <hr />
                                {data.can_edit ?
                                    <>
                                        <EditableTextField defaultContent={data.author.bio} placeholder="Bio" onEdit={setEdited} updateValue={setBio} />
                                        {isEdited &&
                                            <form onSubmit={mutateData.mutate}>
                                                <input type="hidden" name="avatar" value={avatar} />
                                                <input type="hidden" name="display_name" value={displayName} />
                                                <input type="hidden" name="bio" value={bio} />
                                                <input type="submit" value="Save" className="btn btn-primary"></input>
                                            </form>}
                                    </> :
                                    <p>{data.author.bio}</p>
                                }
                            </div>
                        </div>
                        <div className="overflow-scroll">
                            {data.posts.map(post => <CompactPostView key={post.slug} title={post.title} slug={post.slug} author={data.author} />)}
                        </div>
                    </div>
                </div>
            </div>
            {data.can_edit && <WriteButton />}
        </>
    );
}