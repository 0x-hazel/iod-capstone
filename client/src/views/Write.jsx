import { useCallback, useState } from "react";
import useEditView from "../hooks/editView";
import { useNavigate } from "react-router";
import { useMutation } from "@tanstack/react-query";
import axios from "axios";
import qs from "qs";
import MarkdownView from "../components/markdownView";
import NavBar from "../components/navBar";
import { useAlert } from "../hooks/alerts";

export default function Write() {
    const navigate = useNavigate();
    const [title, setTitle] = useState("");
    const [contents, setContents] = useState("");
    const [view, checked] = useEditView();
    const [height, setHeight] = useState(64);
    const alert = useAlert()

    const savePost = useMutation({
        mutationFn: (event) => {
            event.preventDefault();
            const formData = new FormData(event.target);
            return axios.post("/api/post/create-post", qs.stringify(Object.fromEntries(formData.entries())));
        },
        onSuccess: (result) => {
            const path = result?.data?.message;
            if (!path) {
                alert({ status: "error", message: "Failed to navigate to created post." })
            } else {
                navigate(path);
            }
        }
    });

    const updateText = useCallback((event) => {
        setContents(event.target.value);
        setHeight(event.target.scrollHeight);
    }, [setContents, setHeight]);
    return (
        <>
            <NavBar />
            <div className="flex items-center justify-center w-screen flex-col">
                <div className="p-4 w-full md:w-11/17">
                    {view}
                    <div className="mb-4 md:mb-8">
                        {!checked ?
                            <>
                                <input type="text" placeholder="Post Title" className="input text-lg w-full mb-2 md:mb-4" value={title} onChange={(e) => setTitle(e.target.value)} />
                                <textarea className="textarea textarea-secondary w-full" style={{ height: `${height}px` }} value={contents} onChange={updateText}></textarea>
                            </>
                            :
                            <>
                                <h1 className="text-lg my-2 md:my-4 mx-8">{title}</h1>
                                <MarkdownView contents={contents} height={height} />
                            </>
                        }
                    </div>
                    <div className="flex flex-row gap-4">
                        <div className="flex-1"></div>
                        <form className="flex gap-4" onSubmit={savePost.mutate}>
                            <input type="hidden" name="title" value={title} />
                            <input type="hidden" name="content" value={contents} />
                            <button className="btn" onClick={(e) => { e.preventDefault(); navigate(-1) }}>Cancel</button>
                            <button className="btn btn-primary">Save</button>
                        </form>
                    </div>
                </div>
            </div>
        </>
    );
}