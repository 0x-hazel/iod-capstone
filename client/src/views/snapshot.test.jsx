import { render } from "@testing-library/react";
import { expect, test } from "vitest";
import Index from "./Index";
import Login from "./Login";
import Post from "./Post";
import Register from "./Register";
import User from "./User";
import Write from "./Write";
import TestHarness from "../testUtil";

test("snapshot:Index", () => {
    expect(render(
        <TestHarness>
            <Index />
        </TestHarness>
    ).asFragment()).toMatchSnapshot();
});

test("snapshot:Login", () => {
    expect(render(
        <TestHarness>
            <Login />
        </TestHarness>
    ).asFragment()).toMatchSnapshot();
});

test("snapshot:Post", () => {
    expect(render(
        <TestHarness>
            <Post />
        </TestHarness>
    ).asFragment()).toMatchSnapshot();
});

test("snapshot:Register", () => {
    expect(render(
        <TestHarness>
            <Register />
        </TestHarness>
    ).asFragment()).toMatchSnapshot();
});

test("snapshot:User", () => {
    expect(render(
        <TestHarness>
            <User />
        </TestHarness>
    ).asFragment()).toMatchSnapshot();
});

test("snapshot:Write", () => {
    expect(render(
        <TestHarness>
            <Write />
        </TestHarness>
    ).asFragment()).toMatchSnapshot();
});