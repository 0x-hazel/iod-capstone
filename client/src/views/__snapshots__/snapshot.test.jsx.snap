// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot:Index 1`] = `
<DocumentFragment>
  <div
    class="sticky top-0 z-100"
  >
    <div
      class="navbar bg-base-100 shadow-lg"
    >
      <div
        class="flex-1"
      >
        <a
          class="btn btn-ghost text-xl"
          data-discover="true"
          href="/"
        >
          Libris
        </a>
      </div>
      <div
        class="flex-none"
      >
        <ul
          class="menu menu-horizontal px-1"
        >
          <li>
            <a
              data-discover="true"
              href="/login"
            >
              Log In
            </a>
          </li>
          <li>
            <a
              data-discover="true"
              href="/register"
            >
              Register
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-center w-screen flex-col"
  >
    <div
      class="p-4 w-full md:w-11/17"
    >
      <span
        class="loading loading-ring loading-xl"
      />
    </div>
  </div>
  <p>
    Hello, world!
  </p>
</DocumentFragment>
`;

exports[`snapshot:Login 1`] = `
<DocumentFragment>
  <div
    class="sticky top-0 z-100"
  >
    <div
      class="navbar bg-base-100 shadow-lg"
    >
      <div
        class="flex-1"
      >
        <a
          class="btn btn-ghost text-xl"
          data-discover="true"
          href="/"
        >
          Libris
        </a>
      </div>
      <div
        class="flex-none"
      >
        <ul
          class="menu menu-horizontal px-1"
        >
          <li>
            <a
              data-discover="true"
              href="/login"
            >
              Log In
            </a>
          </li>
          <li>
            <a
              data-discover="true"
              href="/register"
            >
              Register
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-center h-screen"
  >
    <div
      class="card card-border bg-base-100 w-96"
    >
      <div
        class="card-body items-center text-center"
      >
        <h2
          class="card-title mb-4"
        >
          Log In
        </h2>
        <form
          class="w-full"
        >
          <label
            class="floating-label mb-8 w-full"
          >
            <span>
              Your Username
            </span>
            <input
              class="input validator"
              name="username"
              placeholder="Username"
              required=""
              type="text"
            />
          </label>
          <label
            class="floating-label mb-8 w-full"
          >
            <span>
              Your Password
            </span>
            <input
              class="input"
              minlength="6"
              name="password"
              placeholder="Password"
              required=""
              type="password"
            />
          </label>
          <div
            class="card-actions"
          >
            <button
              class="btn btn-primary"
            >
              Log In
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`snapshot:Post 1`] = `
<DocumentFragment>
  <p>
    Loading...
  </p>
</DocumentFragment>
`;

exports[`snapshot:Register 1`] = `
<DocumentFragment>
  <div
    class="sticky top-0 z-100"
  >
    <div
      class="navbar bg-base-100 shadow-lg"
    >
      <div
        class="flex-1"
      >
        <a
          class="btn btn-ghost text-xl"
          data-discover="true"
          href="/"
        >
          Libris
        </a>
      </div>
      <div
        class="flex-none"
      >
        <ul
          class="menu menu-horizontal px-1"
        >
          <li>
            <a
              data-discover="true"
              href="/login"
            >
              Log In
            </a>
          </li>
          <li>
            <a
              data-discover="true"
              href="/register"
            >
              Register
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-center h-screen"
  >
    <div
      class="card card-border bg-base-100 w-96"
    >
      <div
        class="card-body items-center text-center"
      >
        <h2
          class="card-title mb-8"
        >
          Register
        </h2>
        <form
          class="w-full"
        >
          <label
            class="floating-label w-full h-24"
          >
            <span>
              Your Username
            </span>
            <input
              class="input validator"
              id="username"
              maxlength="30"
              minlength="3"
              name="username"
              pattern="[A-Za-z][A-Za-z0-9\\-]*"
              placeholder="Username"
              required=""
              title="Only letters, numbers or dash"
              type="text"
            />
            <p
              class="validator-hint"
            >
              Must be 3 to 30 characters containing only letters, numbers or dash.
            </p>
          </label>
          <label
            class="floating-label w-full h-24"
          >
            <span>
              Your Password
            </span>
            <input
              class="input validator"
              id="password"
              minlength="6"
              name="password"
              pattern="(?=.*\\d)(?=.*[a-z])(?=.*[A-Z]).{5,}"
              placeholder="Password"
              required=""
              title="Must be at least 6 characters, including number, lowercase letter, uppercase letter"
              type="password"
              value=""
            />
            <p
              class="validator-hint"
            >
              Must be at least 6 characters, including at least one number, at least one lowercase letter, and at least one uppercase letter.
            </p>
          </label>
          <label
            class="floating-label w-full h-24"
          >
            <span>
              Your Password Again
            </span>
            <input
              class="input validator"
              id="password2"
              minlength="6"
              name="password2"
              pattern=""
              placeholder="Confirm Password"
              required=""
              title="Must match the previous password"
              type="password"
            />
            <p
              class="validator-hint"
            >
              Must match the previous password.
            </p>
          </label>
          <div
            class="card-actions"
          >
            <button
              class="btn btn-primary"
            >
              Register
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`snapshot:User 1`] = `
<DocumentFragment>
  <p>
    Loading...
  </p>
</DocumentFragment>
`;

exports[`snapshot:Write 1`] = `
<DocumentFragment>
  <div
    class="sticky top-0 z-100"
  >
    <div
      class="navbar bg-base-100 shadow-lg"
    >
      <div
        class="flex-1"
      >
        <a
          class="btn btn-ghost text-xl"
          data-discover="true"
          href="/"
        >
          Libris
        </a>
      </div>
      <div
        class="flex-none"
      >
        <ul
          class="menu menu-horizontal px-1"
        >
          <li>
            <a
              data-discover="true"
              href="/login"
            >
              Log In
            </a>
          </li>
          <li>
            <a
              data-discover="true"
              href="/register"
            >
              Register
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-center w-screen flex-col"
  >
    <div
      class="p-4 w-full md:w-11/17"
    >
      <div
        class="grid grid-cols-3 mb-4 md:mb-8"
      >
        <p
          class="text-primary text-right"
        >
          Write
        </p>
        <div
          class="flex items-center justify-center"
        >
          <input
            class="toggle border-primary text-primary checked:border-accent checked:text-accent"
            type="checkbox"
          />
        </div>
        <p />
      </div>
      <div
        class="mb-4 md:mb-8"
      >
        <input
          class="input text-lg w-full mb-2 md:mb-4"
          placeholder="Post Title"
          type="text"
          value=""
        />
        <textarea
          class="textarea textarea-secondary w-full"
          style="height: 64px;"
        />
      </div>
      <div
        class="flex flex-row gap-4"
      >
        <div
          class="flex-1"
        />
        <form
          class="flex gap-4"
        >
          <input
            name="title"
            type="hidden"
            value=""
          />
          <input
            name="content"
            type="hidden"
            value=""
          />
          <button
            class="btn"
          >
            Cancel
          </button>
          <button
            class="btn btn-primary"
          >
            Save
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
